4-bit pte scheme to support all states of existing revoker software (#933)
riscv/riscv-cheri: src/cheri/cheri-pte-ext.adoc

Applied to kernel pages, they can similarly be used against use-after-reallocation within the kernel itself.

The 4-bit _pte_._{rv64y_pte4_field_name_lc}_  field is sub-divided into four 1-bit fields, two controlling capability loads (including the capability loaded during an AMO) and two controlling capability stores (again, including the capability stored as part of an AMO).

[[svycrg_rvy_field_breakdown]]
.Breakdown of the 4-bit _pte_._{rv64y_pte4_field_name_lc}_ field.
[%autowidth,float="center",align="center",cols="<,<,<",options="header"]
|===
| Bit                                  | Name                                | Comment
3+| *Capability load/AMO fields*
|_pte_._{rv64y_pte4_field_name_lc}[0]_ | _pte_._{pte_cap_r_name_lc}_         | Capability readable
|_pte_._{rv64y_pte4_field_name_lc}[1]_ | _pte_._{pte_cap_crg_pte_name_lc}_   | Capability read generation
3+| *Capability store/AMO fields*
|_pte_._{rv64y_pte4_field_name_lc}[2]_ | _pte_._{pte_cap_w_name_lc}_         | Capability writable
|_pte_._{rv64y_pte4_field_name_lc}[3]_ | _pte_._{pte_cap_dirty_name_lc}_     | Capability dirty
|===

> NOTE: {cheri_priv_crg_ext} is a separate extension and not part of the {cheri_base64_ext_name} Privileged ISA, as a future extension may specify an alternative usage for the _pte_._{rv64y_pte4_field_name_lc}_ field allowing a different mechanism giving improved software efficiency.

Reviewer: ```suggestion
NOTE: {cheri_priv_crg_ext} is a separate extension and not part of the {cheri_base64_ext_name} Privileged ISA, to allow future extensions that are incompatible with {cheri_priv_crg_ext} (but compatible with {cheri_base64_ext_name}) to re-purpose these bits.
```

[[pte_rvy_store_summary]]
.Summary of capability store _pte_._{pte_cap_w_name_lc}_ and _pte_._{pte_cap_dirty_name_lc}_ behavior in the PTEs
[%autowidth,float="center",align="center",cols="<,<,<,<",options="header"]
|===
|_pte_._{pte_cap_w_name_lc}_|_pte_._{pte_cap_dirty_name_lc}_|tag^1^| Store Capability Behavior
| 0        | X      |  0   | Normal operation
| 0        | X      |  1   | {cheri_excep_name_pte_st}
| 1        | 0      |  0   | Normal operation
| 1        | 0      |  1   | {cheri_excep_name_pte_st} (_Svade_) or +
hardware _pte_._{pte_cap_dirty_name_lc}_ update (_Svadu_)
| 1        | 1      |  X   | Normal operation
|===

^1^ The to-be-stored {ctag}.

The capability dirty tracking is resolved during memory translation, but there are cases it is not known if there will be a {ctag} stored to memory or not at this point.
Capability dirty tracking _must_ be triggered when _pte_._{pte_cap_w_name_lc}_=1 and _pte_._{pte_cap_dirty_name_lc}_=0 and the to-be-stored {ctag} is set, and _may_ be triggered in the following case where it is not known if there will be a stored {ctag} during translation:

> * SC{LD_ST_DOT_CAP} triggers capability dirty tracking if the {ctag} is set in {cs2}, even if the store fails.

Reviewer: Is this the same behaviour as sc has with the normal dirty bit? If so, how about:
```suggestion
* SC{LD_ST_DOT_CAP} triggers capability dirty tracking if the {ctag} is set in {cs2}, even if the store fails.
  This matches the semantics of SC.* with regard to the _pte_._d_.
```

[NOTE]
====

_pte_._{pte_cap_dirty_name_lc}_ must always be set when there is a stored {ctag} to a virtual page so that software knows which pages have had capabilities stored to them.
It may be set too often, which may cause software to examine the page to check for capabilities when none are present.
This is a situation software is required to handle anyway, as it is always possible for all capabilities in a page to be overwritten by non-capability data.

> In this case the _pte_._{pte_cap_dirty_name_lc}_ bit will still be set.

Reviewer: ```suggestion
In this case the _pte_._{pte_cap_dirty_name_lc}_ bit would still be set.
```

Ideally all implementations would trap precisely (taking the {ctag} into account in all cases) rather than conservatively (trapping more often, potentially on every loaded capability).
However, the loaded {ctag} may not be available in all implementations when determining whether to raise the exception, and therefore flexibility is permitted.
As a result, the software is required to be tolerant of raising the trap when the {ctag} is not set, potentially resulting in spurious traps from pages which have _pte.lyen=1_, and so are likely to store valid capabilities.

Implementations which already take synchronous traps on loaded data, such as ECC faults, are recommended to check the loaded {ctag} when determining whether to raise the fault.

====

[[pte_lyen_lycrg_load_summary]]
.Summary of capability load _pte.lyen_ and _pte.lycrg_ behavior in the PTEs
[%autowidth,float="center",align="center",cols="<,<,<,<,<",options="header"]
|===
|_pte.lyen_| _pte.lycrg_         | _pte.u_|tag^1^| Load Capability Behavior
| 0        | 0                   | X      |  X   | Clear loaded {ctag}
| 0        | 1                   | X      |  X   | Normal operation

> | 1        | X                   | 0      |  X   | Normal operation^2^

Reviewer 2: I suspect that taking into account kernel revocation, it becomes easier to spell out the requirement in the `Load Capability Behavior` predicated on `sstatus.xCRG` with x meaning either UCRG or SCRG:

-  `pte.lyen = 1`, `pte.lycrg == sstatus.xCRG`, `tag = 0`: implementation defined load fault or normal op
-  `pte.lyen = 1`, `pte.lycrg == sstatus.xCRG`, `tag = 1`: normal op
-  `pte.lyen = 1`, `pte.lycrg != sstatus.xCRG`, `tag = 1`: load fault

Unless we want extra reserved states, I'd take out pte.U from the table in this manner.

Such facilities have been shown to provide useful primitives for system-level implementation of certain forms of capability revocation which, in turn, allow userspace heap allocator software to deterministically guard against use-after-reallocation cite:[cornucopia-reloaded].
Additionally, the extension allows tracking of kernel space pages for guarding against use-after-reallocation within the OS kernel.

The 4-bit _pte_._{cheri_base_ext_name_lc}_  field is sub-divided into four 1-bit fields, two controlling capability loads (including the capability loaded during an AMO) and two controlling capability stores (again, including the capability stored as part of an AMO).

[[svycrg_rvy_field_breakdown]]
.Breakdown of the 4-bit _pte_._{cheri_base_ext_name_lc}_ field.
[%autowidth,float="center",align="center",cols="<,<,<",options="header"]
|===
| Bit                                | Name           | Comment
3+| *Capability load/AMO fields*
|_pte_._{cheri_base_ext_name_lc}[0]_ | _pte_._lye_    | Load capability enable
|_pte_._{cheri_base_ext_name_lc}[1]_ | _pte_._lycrg_  | Load capability read generation
3+| *Capability store/AMO fields*

> |_pte_._{cheri_base_ext_name_lc}[2]_ | _pte_._sye_    | Store capability enable

Reviewer 3: CW or YW

Author: YW

|_pte_._{cheri_base_ext_name_lc}[3]_ | _pte_._yd_     | Capability dirty bit
|===

NOTE: {cheri_priv_crg_ext} is not a mandatory part of the {cheri_base64_ext_name} Privileged ISA, as a future extension may specify an alternative usage for the _pte_._{cheri_base_ext_name_lc}_ field allowing a different mechanism giving improved software efficiency.

=== Software Primitives

When all of the following are true, {cheri_priv_crg_ext} adds two primitives for use by software:

. `sstatus.CRGE` is set.
. The authorizing capability has <<c_perm>>.
. The PMA is set to (_CHERI {ctag_title}_).

==== {cheri_excep_name_pte_ld}s

_pte_._lye_ is used to enable capability loads or AMOs to write a {ctag} to {cd}.

> When _pte_._lye_ is set, _pte_._lycrg_ can be used to trap on capability loads or AMOs when it does not match the Capability Read Generation (CRG) value which is represented by the value of `sstatus.UCRG` for userspace pages (_pte_._u=1_) or `sstatus.SCRG` for kernel pages (_pte_._u=0_).

Author: UYRG

The implementation raises a {cheri_excep_name_pte_ld} when, in addition to the rules above, all of the following are true:

. A capability load or AMO is executed.
. _pte_._lye_ is set.
. if _pte_._u=1_ _pte_._lycrg_ does not equal `sstatus.UCRG`.

> . if _pte_._u=0_ _pte_._lycrg_ does not equal `sstatus.SCRG`.

Reviewer 3: sstatus.S[YC]RG matching the PTE bit name

[#section_cheri_priv_crg_ext]
== "{cheri_priv_crg_ext}" Extension, Version 1.0 for {cheri_base64_ext_name}

This extension specifies use of the {rv64y_pte4_field_name_lc} field from the PTE formats
ifndef::cheri_standalone_spec[]
(e.g., see <<sv39pte>>)
endif::[]
.
{cheri_priv_crg_ext} implies Sv39.

The {cheri_priv_crg_ext} extension is enabled when the `sstatus.{revocation_enable}` bit is set.

> NOTE: The enable bit is required to redefine the usage of the _pte_._{rv64y_pte4_field_name_lc}_ field.

Reviewer 2: Is it worth spelling out also future-proofing if other schemes come into existence?

[#section_cheri_priv_crg_ext]
== "{cheri_priv_crg_ext}" Extension, Version 1.0 for {cheri_base64_ext_name}

This extension specifies use of the {rv64y_pte4_field_name_lc} field from the PTE formats (e.g., see <<sv39pte>>).

{cheri_priv_crg_ext} implies Sv39.

The {cheri_priv_crg_ext} extension is enabled when the `sstatus.{revocation_enable}` bit is set.

NOTE: The enable bit is required to redefine the usage of the _pte_._{rv64y_pte4_field_name_lc}_ field.

The extension adds the ability for supervisor-mode software to quickly front-run loads of capabilities from userspace pages, and to track stores of capabilities to pages.
Such facilities have been shown to provide useful primitives for system-level implementation of certain forms of capability revocation which, in turn, allow userspace heap allocator software to deterministically guard against use-after-reallocation cite:[cornucopia-reloaded].

> Additionally, the extension allows tracking of kernel space pages for guarding against use-after-reallocation within the OS kernel.

Reviewer 4: Since sstatus.SYRG is defined, it's not just userspace pages any more?

```suggestion
The extension adds the ability for supervisor-mode software to quickly front-run loads of capabilities from pages and to track stores of capabilities to pages.
Applied to userspace pages, such facilities have been shown to be useful primitives for system-level implementation of certain forms of capability revocation which, in turn, allow userspace heap allocator software to deterministically guard against use-after-reallocation cite:[cornucopia-reloaded].
Applied to kernel pages, they are analogously useful within an OS kernel.
```

[#section_cheri_priv_crg_ext]
== "{cheri_priv_crg_ext}" Extension, Version 1.0 for {cheri_base64_ext_name}

This extension specifies use of the {rv64y_pte4_field_name_lc} field from the PTE formats (e.g., see <<sv39pte>>).

{cheri_priv_crg_ext} implies Sv39.

The {cheri_priv_crg_ext} extension is enabled when the `sstatus.{pte_cap_fc_enable}` bit is set.

NOTE: The enable bit is required to redefine the usage of the _pte_._{rv64y_pte4_field_name_lc}_ field.

> The extension adds the ability for supervisor-mode software to quickly front-run loads of capabilities from pages and to track stores of capabilities to pages.

Reviewer 3: Unless you know how CHERI's load-side revocation works I'm not sure it's obvious what "quickly front-run loads of capabilities" means. Perhaps:

The extension adds the ability for supervisor-mode software to quickly enable trapping loads of capabilities from all pages of memory, incrementally allow loads of capabilities from such pages, and track stores of capabilities to all pages of memory.

> Applied to userspace pages, such facilities have been shown to be useful primitives for system-level implementation of certain forms of capability revocation which, in turn, allow userspace heap allocator software to deterministically guard against use-after-reallocation cite:[cornucopia-reloaded].

Reviewer 4: Sure.  Mostly I've been using "heap" as a short-hand for "not on-stack"; I agree the terminology is terrible.  I'd prefer "certain capability revocation schemes" or similar, since, as stated up thread, the global revocation implemented to date is really not what capability-systems people mean by "revocation".

> Applied to kernel pages, they are analogously useful within an OS kernel.

Reviewer 3: ```suggestion
Applied to kernel pages, they can similarly be used against use-after-reallocation within the kernel itself.
```
since otherwise the sentence seems a bit close to "used for the kernel you can instead use it for the kernel"

The 4-bit _pte_._{rv64y_pte4_field_name_lc}_  field is sub-divided into four 1-bit fields, two controlling capability loads (including the capability loaded during an AMO) and two controlling capability stores (again, including the capability stored as part of an AMO).

[[svycrg_rvy_field_breakdown]]
.Breakdown of the 4-bit _pte_._{rv64y_pte4_field_name_lc}_ field.
[%autowidth,float="center",align="center",cols="<,<,<",options="header"]
|===
| Bit                                  | Name                                | Comment
3+| *Capability load/AMO fields*

> |_pte_._{rv64y_pte4_field_name_lc}[0]_ | _pte_._{pte_cap_r_name_lc}_         | Capability read enable

Author: done

|_pte_._{rv64y_pte4_field_name_lc}[1]_ | _pte_._{pte_cap_crg_pte_name_lc}_   | Capability read generation
3+| *Capability store/AMO fields*
|_pte_._{rv64y_pte4_field_name_lc}[2]_ | _pte_._{pte_cap_w_name_lc}_         | Capability write enable

> |_pte_._{rv64y_pte4_field_name_lc}[3]_ | _pte_._{pte_cap_dirty_name_lc}_     | Capability dirty bit

Author: done

|===

NOTE: {cheri_priv_crg_ext} is not a mandatory part of the {cheri_base64_ext_name} Privileged ISA, as a future extension may specify an alternative usage for the _pte_._{rv64y_pte4_field_name_lc}_ field allowing a different mechanism giving improved software efficiency.

> === Software Primitives

Author: ok

_pte_._{pte_cap_r_name_lc}_ is used to enable capability loads or AMOs to write a {ctag} to {cd}.

When _pte_._{pte_cap_r_name_lc}_ is set, _pte_._{pte_cap_crg_pte_name_lc}_ can be used to trap on capability loads or AMOs when it does not match the Capability Read Generation value which is represented by the value of `sstatus.U{pte_cap_crg_pte_name}` for userspace pages (_pte_._u=1_) or `sstatus.S{pte_cap_crg_pte_name}` for kernel pages (_pte_._u=0_).

The implementation raises a {cheri_excep_name_pte_ld} when, in addition to the rules above, all of the following are true:

. A capability load or AMO is executed.
. _pte_._{pte_cap_r_name_lc}_ is set.
. if _pte_._u=1_, _pte_._{pte_cap_crg_pte_name_lc}_ does not equal `sstatus.U{pte_cap_crg_pte_name}`.
. if _pte_._u=0_, _pte_._{pte_cap_crg_pte_name_lc}_ does not equal `sstatus.S{pte_cap_crg_pte_name}`.
. The loaded {ctag} is set^1^.
.. The {cheri_excep_name_pte_ld} _must_ be taken if the loaded {ctag} is set.
.. The {cheri_excep_name_pte_ld} _may_ be taken if the loaded {ctag} is not set.
. Any other platform specific rules have not forced the loaded {ctag} to be clear.

> NOTE: An example of a platform specific rule is a hardware engine clearing {ctag}s in memory after a call to free() to offload the CPU, which may have cleared a {ctag} before the CPU loads it.

Reviewer 4: I suspect this was an attempt to capture CHERIoT's capability load filter -- additional architecture that intermediates the core's view of capability tags in memory -- and accidentally got cross-threaded with the revocation state machine.  We use the term "filter" to reinforce the non-self-healing nature of the load barrier-in-the-GC-sense: it does not induce a RMW.  (Just to say: the CHERIoT revocation state machine, OTOH, does do RMWs against memory, but is not involved in the load capability path.)

In any case, I agree it's a bit of a mishmash of layers and should probably just go away.

That said, a better NOTE to have here might be to point out the tag-clearing behavior of a capability load instruction citing a `R !C` authorizing cap: because the tag in the register file _will_ be clear, regardless of the tag in memory, the exception need not take place.

> ^1^ Allowing flexibility about checking the loaded {ctag} gives a range of valid implementations.

Reviewer 3: Um, is that not true of every single option in the entirety of RISC-V? Otherwise there'd be no point adding the options. This seems odd to call out for this and only this case.

In this case the _pte_._{pte_cap_dirty_name_lc}_ bit will still be set.

Future AMOs fall into this category:

* For future AMOCAS{LD_ST_DOT_CAP}, it is not known whether the store will happen until the load has executed and the compare has been done. Therefore, AMOCAS{LD_ST_DOT_CAP} is likely to trigger capability dirty tracking if the {ctag} is set in {cs2}.
* For future AMOADD{LD_ST_DOT_CAP}, the stored {ctag} depends upon the loaded {ctag} which is not known during translation and also on the execution of the {CADD}.
 Therefore, AMOADD{LD_ST_DOT_CAP}, is likely to always trigger capability dirty tracking.

Checking the stored {ctag} is less of a burden to the implementation than checking the loaded {ctag} for {cheri_excep_name_pte_ld}, which is why checking the loaded {ctag} is optional behavior.
 However, a future extension may reduce the burden further by removing the check on the to-be-stored {ctag}.

====

NOTE: Capability dirty tracking may be used to implement the store-barrier primitive from cite:[cornucopia-reloaded].

==== Minimum support

> The minimum level of PTE support is to set _pte_._{pte_cap_r_name_lc}_=1, _pte_._{pte_cap_dirty_name_lc}_=1, _pte_._{pte_cap_w_name_lc}_=1 and _pte_._{pte_cap_crg_pte_name_lc}_=0 in all PTEs intended for storing capabilities (e.g., private anonymous mappings) and set `sstatus.U{pte_cap_crg_pte_name}`=0 and `sstatus.S{pte_cap_crg_pte_name}`=0 on all harts, which will enable capabilities to be loaded and stored successfully.

Author: turned into a note
